crates/env/src/engine/on_chain/ext.rs

```diff
@@ -604,9 +607,30 @@ pub fn random(subject: &[u8], output: &mut &mut [u8]) {
     extract_from_slice(output, output_len as usize);
 }
 
-pub fn println(content: &str) {
-    let bytes = content.as_bytes();
-    unsafe { sys::seal_println(Ptr32::from_slice(bytes), bytes.len() as u32) }
+/// If debug message recording is disabled in the contracts pallet, this will be set to false
+/// after an initial call in order to prevent the cost of further calls which will have no effect.
+static mut DEBUG_ENABLED: bool = true;
+
+/// Call `seal_debug_message` with the supplied UTF-8 encoded message.
+///
+/// If debug message recording is disabled in the contracts pallet, the first call to will return
+/// a LoggingDisabled error, and further calls will be a no-op to avoid the cost of calling into
+/// the supervisor.
+///
+/// # Safety
+///
+/// Accesses to the static `DEBUG_ENABLED` are safe because we are executing in a single-threaded
+/// environment.
+pub fn debug_message(message: &str) {
+    if unsafe { DEBUG_ENABLED } {
+        let bytes = message.as_bytes();
+        let ret_code = unsafe {
+            sys::seal_debug_message(Ptr32::from_slice(bytes), bytes.len() as u32)
+        };
+        if let Err(Error::LoggingDisabled) = ret_code.into() {
+            unsafe { DEBUG_ENABLED = false }
+        }
+    }
 }
```

Comment:
I would love to see more encapsulation of the global static mut variable. E.g. put it and the function under an inline module and just export the function from the module. This way we not only encode via comments that the `DEBUG_ENABLED` static should not be used outside of the `debug_message` function.

Reply:
Btw.: What are the cons of putting `DEBUG_ENABLED` into a simple `Cell` dropping all the unsafe code? I think `Cell` does not really have a serious overhead, doesn't it?

Reply:
I believe `thread_local!` is only available in `std`

Reply:
> Btw.: What are the cons of putting DEBUG_ENABLED into a simple Cell dropping all the unsafe code? I think Cell does not really have a serious overhead, doesn't it?

`Cell` is not `Sync` and cannot put into a static. You have the choice between unsafe and using a synchronization primitive like a spinlock. The latter ones would of course introduce overhead not necessary in a singlethreaded environment. You could of course newtype wrap a `Cell` and implement `Sync` for it. That should be safe everywhere in ink!.

> I believe thread_local! is only available in std

Correct. The concept of threads does not exist in a vaccuum. It needs help from the operaring system. Which in this case is pallet_contracts. That said,  who mentioned `thread_local`?

Reply:
> That said, who did mention thread_local!?

Ha, yes usually that is a way to use `Cell` in a `static` to avoid the `Sync` requirement.

Reply:
> Putting the static variable inside the function is a good idea as its only referenced from there.

Hadn't thought of that. Done!

Reply:
> That said, who did mention thread_local!?

We use `thread_local!` in ink! for the off-chain environment testing infrastructure.